File: tests/algorithms/x/Java/machine_learning/automatic_differentiation.java
public class Main {
    static class Dual {
        double value;
        double deriv;
        Dual(double value, double deriv) {
            this.value = value;
            this.deriv = deriv;
        }
        Dual() {}
        @Override public String toString() {
            return String.format("{'value': %s, 'deriv': %s}", String.valueOf(value), String.valueOf(deriv));
        }
    }


    static Dual dual(double v, double d) {
        return new Dual(v, d);
    }

    static double pow_float(double base, long exp) {
        double res = 1.0;
        long i_1 = 0L;
        while (i_1 < exp) {
            res = res * base;
            i_1 = i_1 + 1;
        }
        return res;
    }

    static Dual add(Dual a, Dual b) {
        return new Dual(a.value + b.value, a.deriv + b.deriv);
    }

    static Dual sub(Dual a, Dual b) {
        return new Dual(a.value - b.value, a.deriv - b.deriv);
    }

    static Dual mul(Dual a, Dual b) {
        return new Dual(a.value * b.value, a.deriv * b.value + b.deriv * a.value);
    }

    static Dual div(Dual a, Dual b) {
        return new Dual(a.value / b.value, (a.deriv * b.value - b.deriv * a.value) / (b.value * b.value));
    }

    static Dual power(Dual a, long p) {
        return new Dual(pow_float(a.value, p), (1.0 * p) * pow_float(a.value, p - 1) * a.deriv);
    }

    static void main() {
        Dual a = dual(2.0, 1.0);
        Dual b_1 = dual(1.0, 0.0);
        Dual c_1 = add(a, b_1);
        Dual d_1 = mul(a, b_1);
        Dual e_1 = div(c_1, d_1);
        System.out.println(_p(e_1.deriv));
        Dual x_1 = dual(2.0, 1.0);
        Dual y_1 = power(x_1, 3L);
        System.out.println(_p(y_1.deriv));
    }
    public static void main(String[] args) {
        main();
    }

    static String _p(Object v) {
        if (v == null) return "<nil>";
        if (v.getClass().isArray()) {
            if (v instanceof int[]) return java.util.Arrays.toString((int[]) v);
            if (v instanceof long[]) return java.util.Arrays.toString((long[]) v);
            if (v instanceof double[]) return java.util.Arrays.toString((double[]) v);
            if (v instanceof boolean[]) return java.util.Arrays.toString((boolean[]) v);
            if (v instanceof byte[]) return java.util.Arrays.toString((byte[]) v);
            if (v instanceof char[]) return java.util.Arrays.toString((char[]) v);
            if (v instanceof short[]) return java.util.Arrays.toString((short[]) v);
            if (v instanceof float[]) return java.util.Arrays.toString((float[]) v);
            return java.util.Arrays.deepToString((Object[]) v);
        }
        if (v instanceof Double || v instanceof Float) {
            double d = ((Number) v).doubleValue();
            if (d == Math.rint(d)) return String.valueOf((long) d);
            return String.valueOf(d);
        }
        return String.valueOf(v);
    }
}

File: tests/algorithms/transpiler/FS/machine_learning/automatic_differentiation.fs
// Generated 2025-08-08 17:07 +0700

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _dictAdd<'K,'V when 'K : equality> (d:System.Collections.Generic.IDictionary<'K,'V>) (k:'K) (v:'V) =
    d.[k] <- v
    d
let _dictCreate<'K,'V when 'K : equality> (pairs:('K * 'V) list) : System.Collections.Generic.IDictionary<'K,'V> =
    let d = System.Collections.Generic.Dictionary<'K, 'V>()
    for (k, v) in pairs do
        d.[k] <- v
    upcast d
let _dictGet<'K,'V when 'K : equality> (d:System.Collections.Generic.IDictionary<'K,'V>) (k:'K) : 'V =
    match d.TryGetValue(k) with
    | true, v -> v
    | _ -> Unchecked.defaultof<'V>
let rec _str v =
    let s = sprintf "%A" v
    s.Replace("[|", "[")
     .Replace("|]", "]")
     .Replace("; ", " ")
     .Replace(";", "")
     .Replace("\"", "")
type Dual = {
    mutable _value: float
    mutable _deriv: float
}
let rec dual (v: float) (d: float) =
    let mutable __ret : Dual = Unchecked.defaultof<Dual>
    let mutable v = v
    let mutable d = d
    try
        __ret <- { _value = v; _deriv = d }
        raise Return
        __ret
    with
        | Return -> __ret
and pow_float (``base``: float) (exp: int) =
    let mutable __ret : float = Unchecked.defaultof<float>
    let mutable ``base`` = ``base``
    let mutable exp = exp
    try
        let mutable res: float = 1.0
        let mutable i: int = 0
        while i < exp do
            res <- res * ``base``
            i <- i + 1
        __ret <- res
        raise Return
        __ret
    with
        | Return -> __ret
and add (a: Dual) (b: Dual) =
    let mutable __ret : Dual = Unchecked.defaultof<Dual>
    let mutable a = a
    let mutable b = b
    try
        __ret <- { _value = (a._value) + (b._value); _deriv = (a._deriv) + (b._deriv) }
        raise Return
        __ret
    with
        | Return -> __ret
and sub (a: Dual) (b: Dual) =
    let mutable __ret : Dual = Unchecked.defaultof<Dual>
    let mutable a = a
    let mutable b = b
    try
        __ret <- { _value = (a._value) - (b._value); _deriv = (a._deriv) - (b._deriv) }
        raise Return
        __ret
    with
        | Return -> __ret
and mul (a: Dual) (b: Dual) =
    let mutable __ret : Dual = Unchecked.defaultof<Dual>
    let mutable a = a
    let mutable b = b
    try
        __ret <- { _value = (a._value) * (b._value); _deriv = ((a._deriv) * (b._value)) + ((b._deriv) * (a._value)) }
        raise Return
        __ret
    with
        | Return -> __ret
and div (a: Dual) (b: Dual) =
    let mutable __ret : Dual = Unchecked.defaultof<Dual>
    let mutable a = a
    let mutable b = b
    try
        __ret <- { _value = (a._value) / (b._value); _deriv = (((a._deriv) * (b._value)) - ((b._deriv) * (a._value))) / ((b._value) * (b._value)) }
        raise Return
        __ret
    with
        | Return -> __ret
and power (a: Dual) (p: int) =
    let mutable __ret : Dual = Unchecked.defaultof<Dual>
    let mutable a = a
    let mutable p = p
    try
        __ret <- { _value = pow_float (a._value) (p); _deriv = ((1.0 * (float p)) * (pow_float (a._value) (p - 1))) * (a._deriv) }
        raise Return
        __ret
    with
        | Return -> __ret
and main () =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    try
        let __bench_start = _now()
        let __mem_start = System.GC.GetTotalMemory(true)
        let a: Dual = dual (2.0) (1.0)
        let b: Dual = dual (1.0) (0.0)
        let c: Dual = add (a) (b)
        let d: Dual = mul (a) (b)
        let e: Dual = div (c) (d)
        printfn "%s" (_str (e._deriv))
        let x: Dual = dual (2.0) (1.0)
        let y: Dual = power (x) (3)
        printfn "%s" (_str (y._deriv))
        let __bench_end = _now()
        let __mem_end = System.GC.GetTotalMemory(true)
        printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

        __ret
    with
        | Return -> __ret
main()
